### PR TITLE
add callback to override destination host and port for a direct-tcpip…

### PR DESCRIPTION
**Pull request by maigl**
… connection

### CODE REVIEW

File: tcpip.go [hunk 1]
```diff
@@ -24,13 +24,18 @@ func directTcpipHandler(srv *Server, conn *gossh.ServerConn, newChan gossh.NewCh
 		return
 	}
 
-	if srv.LocalPortForwardingCallback == nil || !srv.LocalPortForwardingCallback(ctx, d.DestinationHost, d.DestinationPort) {
+	pdf := &PortForwardDestination{
```
Comment (chressie) [inline tcpip.go]:
```suggestion
	pfd := &PortForwardDestination{
```

File: tcpip.go [hunk 2]
```diff
@@ -24,13 +24,18 @@ func directTcpipHandler(srv *Server, conn *gossh.ServerConn, newChan gossh.NewCh
...
+		d.DestinationHost,
+		int(d.DestinationPort),
+	}
+
+	if srv.LocalPortForwardingCallback == nil || !srv.LocalPortForwardingCallback(ctx, pdf) {
```
Comment (chressie) [inline tcpip.go]:
```suggestion
	if srv.LocalPortForwardingCallback == nil || !srv.LocalPortForwardingCallback(ctx, pfd) {
```

File: tcpip.go [hunk 3]
```diff
@@ -24,13 +24,18 @@ func directTcpipHandler(srv *Server, conn *gossh.ServerConn, newChan gossh.NewCh
...
 		newChan.Reject(gossh.Prohibited, "port forwarding is disabled")
 		return
 	}
 
-	dest := net.JoinHostPort(d.DestinationHost, strconv.FormatInt(int64(d.DestinationPort), 10))
-	
+	dest := net.JoinHostPort(pdf.Host, strconv.Itoa(pdf.Port))
```
Comment (chressie) [inline tcpip.go]:
```suggestion
	dest := net.JoinHostPort(pfd.Host, strconv.Itoa(pfd.Port))
```

File: tcpip_test.go [hunk 4]
```diff
@@ -48,6 +49,34 @@ func newTestSessionWithForwarding(t *testing.T, forwardingEnabled bool) (net.Lis
...
+func newTestSessionWithForwardingAndOverwriteDestination(t *testing.T, forwardingEnabled bool) (net.Listener, *gossh.Client, func()) {
+	l := sampleSocketServer(sampleServerResponse)
+	alternativeDestination := sampleSocketServer(altSampleServerResponse)
+
+	_, client, cleanup := newTestSession(t, &Server{
+		Handler: func(s Session) {},
+		LocalPortForwardingCallback: func(ctx Context, pdf *PortForwardDestination) bool {
+			addr := net.JoinHostPort(pdf.Host, strconv.Itoa(pdf.Port))
+			if addr != l.Addr().String() {
+				panic("unexpected destinationHost: " + addr)
```
Comment (chressie) [inline tcpip_test.go]: panic? maybe t.Fatal

File: tcpip_test.go [hunk 5]
```diff
@@ -48,6 +49,34 @@ func newTestSessionWithForwarding(t *testing.T, forwardingEnabled bool) (net.Lis
...
+			}
+			tmp := strings.Split(alternativeDestination.Addr().String(), ":")
```
Comment (chressie) [inline tcpip_test.go]: net.SplitHostPort

File: tcpip_test.go [hunk 6]
```diff
@@ -67,6 +96,25 @@ func TestLocalPortForwardingWorks(t *testing.T) {
...
+	conn, err := client.Dial("tcp", l.Addr().String())
+	if err != nil {
+		t.Fatalf("Error connecting to %v: %v", l.Addr().String(), err)
+	}
+	result, err := ioutil.ReadAll(conn)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(result, altSampleServerResponse) {
+		t.Fatalf("result = %#v; want %#v", result, sampleServerResponse)
```
Comment (chressie) [inline tcpip_test.go:118]: printing with %s will automatically print the responses as string.

File: ssh.go [hunk 7]
```diff
@@ -47,8 +47,16 @@ type PtyCallback func(ctx Context, pty Pty) bool
 // the net.Conn that will be used as the underlying connection.
 type ConnCallback func(conn net.Conn) net.Conn
 
-// LocalPortForwardingCallback is a hook for allowing port forwarding
-type LocalPortForwardingCallback func(ctx Context, destinationHost string, destinationPort uint32) bool
+// PortForwardDestination holds Host and Port of the port forward destination.
+type PortForwardDestination struct {
+	Host string
+	Port int
+}
```
Comment (chressie) [inline ssh.go:55]: maybe add a func (p *PortForwardDestination) String() string { return net.JoinHostPort(p.Host, strconv.Itoa(p.Port)) }

**Review by chressie — COMMENTED** [review]
A few nits.

**Comment by maigl** [comment]
okay .. I fixed those nits i found useful

**Review by chressie — APPROVED** [review]
LGTM

**Comment by sbueringer** [comment]
Finally! :)

**Pull request merged**